### PR TITLE
feat: support `server.base` configuration

### PR DESCRIPTION
**Pull request by 9aoy**
## Summary

Support modifying the base path of the server through `server.base` configuration. 

By default, the base path of the server is `/`, and users can access output files such as `index.html` and public folder assets through http://localhost:3000/.

When we want to access these files through http://localhost:3000/foo/, we can configure the following:

```ts title="rsbuild.config.ts"
export default {
  server: {
    base: '/foo',
  },
};
```

## Related Links

https://github.com/web-infra-dev/rsbuild/issues/1242

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

**Comment by nanianlisao** [comment]
not so sure about this setup, especially its link with output.assetPrefix. Looks like server and output should be split.

If it is just for processing publicDir, this also needs to affect copyOnBuild. Should you consider placing it in publicDir, such as publicDir.base?

**Comment by 9aoy** [comment]
> not so sure about this setup, especially its link with output.assetPrefix. Looks like server and output should be split.

Configuring `server.base` is usually because not all projects will be deployed to the root path. At this time, `server.base` can be configured as a subpath, which ensures that local development and preview are consistent with the online access path.

And `server.base` is just the default value of `output.assetPrefix`. If `output.assetPrefix` is ​​defined, output.assetPrefix's own configuration will be used.

> If it is just for processing publicDir, this also needs to affect copyOnBuild. Should you consider placing it in publicDir, such as publicDir.base?

Not just for processing public Dir. You can read related discussions https://github.com/web-infra-dev/rsbuild/issues/1242

**Comment by nanianlisao** [comment]
> Configuring server.base is usually because not all projects will be deployed to the root path.

I think this should be something output.assetPrefix controls, and when I have a problem I'll check it first, rather than some config I might be overlooking.

Judging from #1242, I think the core requirement is still that server.publicDir needs to support specifying additional directories when copying.

It's also worth mentioning that I don't like dev.assetPrefix either. Using `process.env.NODE_ENV === 'production' ? '/cv3/' : '/'` makes me less prone to making mistakes.

**Comment by chenjiahan** [comment]
> It's also worth mentioning that I don't like dev.assetPrefix either. Using process.env.NODE_ENV === 'production' ? '/cv3/' : '/' makes me less prone to making mistakes.

I am neutral about `dev.assetPrefix`. We have considered remove it to reduce configuration options, but it will break early users and some ByteDance projects.

In most projects I've encountered, the `assetPrefix` used by dev and prod is usually different, users will use a CDN url as `assetPrefix` in the prod mode. So it still makes sense to provide a separate `assetPrefix` option for dev mode.

File: website/docs/en/guide/basic/static-assets.mdx [hunk 1]
```diff
@@ -112,10 +112,10 @@ For example, you can place files such as `robots.txt`, `manifest.json`, or `favi
 
 You can reference files in the `public` directory via a URL.
 
-For example, in an HTML template, the `./public/favicon.ico` file can be referenced as `/favicon.ico`.
+For example, in an HTML template, the `./public/favicon.ico` file can be referenced as `/favicon.ico`, [BASE_URL](/guide/advanced/env-vars#processenvpublic_base_path) is the base path of the server.
 
 ```html title="index.html"
-<link rel="icon" href="/favicon.ico" />
+<link rel="icon" href="<%= process.env.BASE_URL %>/favicon.ico" />
```
Comment (chenjiahan) [inline website/docs/en/guide/basic/static-assets.mdx:118]: BASE_URL + '/' = double slash?

File: packages/core/src/plugins/define.ts [hunk 2]
```diff
@@ -14,6 +14,9 @@ export const pluginDefine = (): RsbuildPlugin => ({
         'process.env.ASSET_PREFIX': JSON.stringify(
           getPublicPathFromChain(chain, false),
         ),
+        'process.env.BASE_URL': JSON.stringify(
```
Comment (chenjiahan) [inline packages/core/src/plugins/define.ts:17]: Should add `process.env.BASE_URL` to env variables documentation

**Review by chenjiahan — COMMENTED** [review]
Since `base` is a new option, I propose to make it stricter:

1. A `base` value should always end with a trailing slash: '/foo/'.
2. We do not need to remove the tailing slash in `process.env.BASE_URL`

File: website/docs/en/guide/basic/static-assets.mdx (hunk 1 above)
Comment (9aoy) [reply website/docs/en/guide/basic/static-assets.mdx:118]: Will automatically remove the trailing slash from `BASE_URL`.

I prefer to consistent with [process.env.ASSET_PREFIX](https://rsbuild.dev/guide/advanced/env-vars#processenvasset_prefix).

Comment (chenjiahan) [reply website/docs/en/guide/basic/static-assets.mdx:118]: I tend to align with the ecosystem, we can check Next.js / Vite before making a decision.

**Comment by FugitiveLc** [comment]
I think this configuration is necessary because I also encountered this problem when migrating old projects

File: website/docs/en/guide/basic/static-assets.mdx (hunk 1 above)
Comment (9aoy) [reply website/docs/en/guide/basic/static-assets.mdx:118]: Both vite and next.js are output to the front-end env as is according to the `baseUrl` configured by the user.

So, I think we don't need to do anything to `process.env.BASE_URL`, just leave it as is.

Comment (chenjiahan) [reply website/docs/en/guide/basic/static-assets.mdx:118]: LGTM

File: website/docs/en/config/server/base.mdx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+- **Type:** `string`
+- **Default:** `/`
+
+`server.base` is used to configure the [base path](/guide/basic/server#base-path) of the server.
+
+### Example
+
+By default, the base path of the server is `/`, and users can access `index.html` through http://localhost:3000.
+
+When we want to access `index.html` through http://localhost:3000/foo, we can modify `server.base` to `/foo`.
```
Comment (chenjiahan) [inline website/docs/en/config/server/base.mdx:12]:
```suggestion
If you want to access `index.html` through `http://localhost:3000/foo`, you can change `server.base` to `/foo`.
```

File: website/docs/en/config/server/base.mdx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+export default {
+  server: {
+    base: '/foo',
+  },
+};
+```
+
+### The URL prefix of static assets
+
+By default, [dev.assetPrefix](/config/dev/asset-prefix) and [output.assetPrefix](/config/output/asset-prefix) will read the value of `server.base` as the default value of `assetPrefix`.
```
Comment (chenjiahan) [inline website/docs/en/config/server/base.mdx:24]:
```suggestion
By default, [dev.assetPrefix](/config/dev/asset-prefix) and [output.assetPrefix](/config/output/asset-prefix) will read the value of `server.base` as the default value.
```

File: website/docs/en/config/server/base.mdx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+When `server.base` is `/foo`, the default resource URL loaded in the browser is as follows:
+
+```html
+<script defer src="/foo/static/js/index.js"></script>
+```
+
+Then, `index.html` and static asset outputs can be accessed through http://localhost:3000/foo.
+
+If you do not wish to use this default behavior, you can override it by explicitly setting `dev.assetPrefix` / `output.assetPrefix` :
```
Comment (chenjiahan) [inline website/docs/en/config/server/base.mdx:34]:
```suggestion
If you do not want to use this default behavior, you can override it by explicitly setting `dev.assetPrefix` / `output.assetPrefix` :
```

File: website/docs/en/guide/advanced/env-vars.mdx [hunk 6]
```diff
@@ -107,6 +107,28 @@ In production mode, the above code will be compiled to:
...
+
+```ts
+export default {
+  server: {
+    base: '/foo',
+  },
+};
+```
+
+Then, the access path to the `favicon.ico` file in the public directory is `http://localhost:3000/foo/favicon.ico`. We can reference `./public/favicon.ico` file in the HTML template in the following way:
```
Comment (chenjiahan) [inline website/docs/en/guide/advanced/env-vars.mdx:126]:
```suggestion
Then, the access path to the `favicon.ico` file in the public directory is `http://localhost:3000/foo/favicon.ico`. We can reference `./public/favicon.ico` file in the HTML template like this:
```

File: website/docs/zh/config/output/asset-prefix.mdx [hunk 7]
```diff
@@ -30,6 +30,25 @@ export default {
 ></script>
 ```
 
+## 默认值
+
+`output.assetPrefix` 的默认值与 [server.base](/config/server/base) 相同。
+
+当 `server.base` 为 `/foo` 时，可通过 http://localhost:3000/foo 访问到 `index.html` 及静态资源产物。
+
+需要注意的是，当自定义 `output.assetPrefix` 选项时，如果希望静态资源能够通过 Rsbuild 预览服务器正常访问到，`output.assetPrefix` 应和 `server.base` 包含相同的 url 前缀，如：
```
Comment (chenjiahan) [inline website/docs/zh/config/output/asset-prefix.mdx:39]:
```suggestion
需要注意的是，当自定义 `output.assetPrefix` 选项时，如果希望静态资源能够通过 Rsbuild 预览服务器正常访问，`output.assetPrefix` 应和 `server.base` 包含相同的 URL 前缀，如：
```

File: website/docs/zh/config/server/base.mdx [hunk 8]
```diff
@@ -0,0 +1,49 @@
+# server.base
+
+- **类型：** `string`
+
+- **默认值：** `/`
```
Comment (chenjiahan) [inline website/docs/zh/config/server/base.mdx:5]:
```suggestion
- **默认值：** `/`
```

File: website/docs/zh/config/server/base.mdx [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+`server.base` 用于配置服务端的[基础路径](/guide/basic/server#服务端基础路径)。
+
+### 示例
+
+默认情况下，服务端的基础路径为 `/`，用户可通过 http://localhost:3000 访问到 `index.html`。
+
+当我们希望通过 http://localhost:3000/foo 访问到 `index.html` 时，可将 `server.base` 修改为 `/foo`。
```
Comment (chenjiahan) [inline website/docs/zh/config/server/base.mdx:13]:
```suggestion
当你希望通过 `http://localhost:3000/foo/` 访问到 `index.html` 时，可以将 `server.base` 修改为 `/foo`。
```

File: website/docs/zh/config/server/base.mdx [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+默认情况下，[dev.assetPrefix](/config/dev/asset-prefix) 和 [output.assetPrefix](/config/output/asset-prefix) 会读取 `server.base` 的值作为 `assetPrefix` 的默认值。
+
+当 `server.base` 为 `/foo` 时，默认在浏览器中加载的资源 URL 如下：
+
+```html
+<script defer src="/foo/static/js/index.js"></script>
+```
+
+此时，可通过 http://localhost:3000/foo 访问到 `index.html` 及静态资源产物。
```
Comment (chenjiahan) [inline website/docs/zh/config/server/base.mdx:33]:
```suggestion
此时，可通过 `http://localhost:3000/foo/` 访问到 `index.html` 及静态资源产物。
```

File: website/docs/en/config/dev/asset-prefix.mdx [hunk 11]
```diff
@@ -1,14 +1,33 @@
...
 
 `assetPrefix` will affect the URLs of most of the static assets, including JavaScript files, CSS files, images, videos, etc. If an incorrect value is specified, you'll receive 404 errors while loading these resources.
 
 This config is only used in development mode. In the production mode, please use the [output.assetPrefix](/config/output/asset-prefix) to set the URL prefix.
 
+## Default Value
+
+The default value of `dev.assetPrefix` is the same as [server.base](/config/server/base).
+
+When `server.base` is `/foo`, `index.html` and other static assets can be accessed through http://localhost:3000/foo.
```
Comment (chenjiahan) [inline website/docs/en/config/dev/asset-prefix.mdx:16]:
```suggestion
When `server.base` is `/foo`, `index.html` and other static assets can be accessed through `http://localhost:3000/foo/`.
```

File: website/docs/en/config/output/asset-prefix.mdx [hunk 12]
```diff
@@ -30,6 +30,25 @@ After building, you can see that the JS files are loaded from:
 ></script>
 ```
 
+## Default Value
+
+The default value of `output.assetPrefix` is the same as [server.base](/config/server/base).
+
+When `server.base` is `/foo`, `index.html` and static assets can be accessed through http://localhost:3000/foo.
```
Comment (chenjiahan) [inline website/docs/en/config/output/asset-prefix.mdx:37]:
```suggestion
When `server.base` is `/foo`, `index.html` and static assets can be accessed through `http://localhost:3000/foo/`.
```

File: website/docs/en/config/server/base.mdx [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+By default, [dev.assetPrefix](/config/dev/asset-prefix) and [output.assetPrefix](/config/output/asset-prefix) will read the value of `server.base` as the default value of `assetPrefix`.
+
+When `server.base` is `/foo`, the default resource URL loaded in the browser is as follows:
+
+```html
+<script defer src="/foo/static/js/index.js"></script>
+```
+
+Then, `index.html` and static asset outputs can be accessed through http://localhost:3000/foo.
```
Comment (chenjiahan) [inline website/docs/en/config/server/base.mdx:32]:
```suggestion
Then, `index.html` and static asset outputs can be accessed through `http://localhost:3000/foo/`.
```

File: website/docs/en/config/server/base.mdx [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+- **Type:** `string`
+- **Default:** `/`
+
+`server.base` is used to configure the [base path](/guide/basic/server#base-path) of the server.
+
+### Example
+
+By default, the base path of the server is `/`, and users can access `index.html` through http://localhost:3000.
+
+If you want to access `index.html` through `http://localhost:3000/foo`, you can change `server.base` to `/foo`.
```
Comment (chenjiahan) [inline website/docs/en/config/server/base.mdx:12]:
```suggestion
If you want to access `index.html` through `http://localhost:3000/foo/`, you can change `server.base` to `/foo`.
```

File: website/docs/en/guide/basic/server.mdx [hunk 15]
```diff
@@ -2,16 +2,32 @@
 
 Rsbuild comes with a built-in dev server designed to improve the development experience. When you run the `rsbuild dev` or `rsbuild preview` commands, the server will start, providing features such as page preview, routing, and hot module reloading.
 
+## Base Path
+
+By default, the base path of the server is `/`, and users can access output files such as `index.html` and public folder assets through http://localhost:3000/.
+
+Rsbuild supports modifying the base path of the server through [server.base](/config/server/base). When we want to access these files through http://localhost:3000/foo/, we can configure the following:
```
Comment (chenjiahan) [inline website/docs/en/guide/basic/server.mdx:9]:
```suggestion
Rsbuild supports modifying the base path of the server through [server.base](/config/server/base). If you to access these files through `http://localhost:3000/foo/`, you can configure the following:
```

File: website/docs/zh/config/dev/asset-prefix.mdx [hunk 16]
```diff
@@ -1,14 +1,33 @@
...
 
 `assetPrefix` 会影响构建产物中绝大部分静态资源的 URL，包括 JavaScript 文件、CSS 文件、图片、视频等。如果指定了一个错误的值，则在加载这些资源时可能会出现 404 错误。
 
 该配置项仅用于开发模式。在生产模式下，请使用 [output.assetPrefix](/config/output/asset-prefix) 配置项进行设置。
 
+## 默认值
+
+`dev.assetPrefix` 的默认值与 [server.base](/config/server/base) 相同。
+
+当 `server.base` 为 `/foo` 时，可通过 http://localhost:3000/foo 访问到 `index.html` 及其他静态资源产物。
```
Comment (chenjiahan) [inline website/docs/zh/config/dev/asset-prefix.mdx:16]:
```suggestion
当 `server.base` 为 `/foo` 时，可通过 `http://localhost:3000/foo/` 访问到 `index.html` 及其他静态资源产物。
```

File: website/docs/zh/config/output/asset-prefix.mdx [hunk 17]
```diff
@@ -30,6 +30,25 @@ export default {
 ></script>
 ```
 
+## 默认值
+
+`output.assetPrefix` 的默认值与 [server.base](/config/server/base) 相同。
+
+当 `server.base` 为 `/foo` 时，可通过 http://localhost:3000/foo 访问到 `index.html` 及静态资源产物。
```
Comment (chenjiahan) [inline website/docs/zh/config/output/asset-prefix.mdx:37]:
```suggestion
当 `server.base` 为 `/foo` 时，可通过 `http://localhost:3000/foo/` 访问到 `index.html` 及静态资源产物。
```

File: website/docs/zh/config/server/base.mdx [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+默认情况下，[dev.assetPrefix](/config/dev/asset-prefix) 和 [output.assetPrefix](/config/output/asset-prefix) 会读取 `server.base` 的值作为 `assetPrefix` 的默认值。
+
+当 `server.base` 为 `/foo` 时，默认在浏览器中加载的资源 URL 如下：
+
+```html
+<script defer src="/foo/static/js/index.js"></script>
+```
+
+此时，可通过 `http://localhost:3000/foo/` 访问到 `index.html` 及静态资源产物。
```
Comment (chenjiahan) [inline website/docs/zh/config/server/base.mdx:32]:
```suggestion
此时，可以通过 `http://localhost:3000/foo/` 访问到 `index.html` 以及其他静态资源产物。
```

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**